# Project name!

<span style='background:mistyrose'>&nbsp;</span>&nbsp;

## About the dataset

#### Source:
[Wine quality dataset](https://archive.ics.uci.edu/ml/datasets/wine+quality)

Two datasets are included, related to red and white wine samples from the demarcated region of *vinho verde* (northeast region of Portugal). Due to privacy and logistic issues, only physicochemical (inputs) and sensory (the output) variables are available (e.g. there is no data about grape types, wine brand, wine selling price, etc.). 

Each sample was evaluated by a minimum of three sensory assessors (using blind tastes), which graded the wine in a scale that ranges from 0 (very bad) to 10 (excellent). The final sensory score is given by the *median* of these evaluations.

#### Attribute Information:    
##### &ensp;- Input variables (based on physicochemical tests):    
&ensp;`fixed acidity` (tartaric acid - g$\cdot$dm<sup>-3</sup>):<br>
&ensp;&ensp;acidity is a fundamental property of wine, imparting sourness and resistance to microbial infection<br>
&ensp;`volatile acidity` (acetic acid - g$\cdot$dm<sup>-3</sup>):<br>
&ensp;&ensp;Wine spoilage is legally defined by volatile acidity, largely composed of acetic acid<br>
&ensp;`citric acid` (g$\cdot$dm<sup>-3</sup>)<br>
&ensp;&ensp;description<br>
&ensp;`residual sugar` (g$\cdot$dm<sup>-3</sup>)<br>
&ensp;&ensp;description<br>
&ensp;`chlorides` (sodium chloride - g$\cdot$dm<sup>-3</sup>)<br>
&ensp;&ensp;description<br>
&ensp;`free sulfur dioxide` (mg$\cdot$dm<sup>-3</sup>)<br>
&ensp;&ensp;description<br>
&ensp;`total sulfur dioxide` (mg$\cdot$dm<sup>-3</sup>)<br>
&ensp;&ensp;description<br>
&ensp;`density` (g$\cdot$dm<sup>-3</sup>)<br>
&ensp;&ensp;description<br>
&ensp;`pH`:<br>
&ensp;&ensp;Typical pH levels in wine normally range from 2.9 to 3.9<br>
&ensp;`sulphates` (potassium sulphate - g$\cdot$dm<sup>-3</sup>)<br>
&ensp;&ensp;description<br>
&ensp;`alcohol` (% by volume)<br>
&ensp;&ensp;description

##### &ensp;- Output variable (based on sensory data):    
&ensp;&ensp;`quality` (score between 0 and 10)


#### Main goal:
Model wine quality based on physicochemical tests

## Import libraries

In [1]:
import pandas as pd
import seaborn as sns

%config Completer.use_jedi = False

## Load the data

In [2]:
red_wine = pd.read_csv('Wine quality/winequality-red.csv', sep=";")
white_wine = pd.read_csv('Wine quality/winequality-white.csv', sep=";")

# Add a tag to identify each type on a combined dataset
red_wine['type'] = 'red'
white_wine['type'] = 'white'

In [3]:
red_wine.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1599 entries, 0 to 1598
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         1599 non-null   float64
 1   volatile acidity      1599 non-null   float64
 2   citric acid           1599 non-null   float64
 3   residual sugar        1599 non-null   float64
 4   chlorides             1599 non-null   float64
 5   free sulfur dioxide   1599 non-null   float64
 6   total sulfur dioxide  1599 non-null   float64
 7   density               1599 non-null   float64
 8   pH                    1599 non-null   float64
 9   sulphates             1599 non-null   float64
 10  alcohol               1599 non-null   float64
 11  quality               1599 non-null   int64  
 12  type                  1599 non-null   object 
dtypes: float64(11), int64(1), object(1)
memory usage: 162.5+ KB


In [4]:
white_wine.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4898 entries, 0 to 4897
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         4898 non-null   float64
 1   volatile acidity      4898 non-null   float64
 2   citric acid           4898 non-null   float64
 3   residual sugar        4898 non-null   float64
 4   chlorides             4898 non-null   float64
 5   free sulfur dioxide   4898 non-null   float64
 6   total sulfur dioxide  4898 non-null   float64
 7   density               4898 non-null   float64
 8   pH                    4898 non-null   float64
 9   sulphates             4898 non-null   float64
 10  alcohol               4898 non-null   float64
 11  quality               4898 non-null   int64  
 12  type                  4898 non-null   object 
dtypes: float64(11), int64(1), object(1)
memory usage: 497.6+ KB


Both datasets have consistent columns, all columns are numerical (except for `type` that we just added) and there are no missing values. Let's concatenate them in a single dataframe.

In [5]:
wine_df = pd.concat([red_wine, white_wine]).reset_index(drop=True)
wine_df

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0               7.4              0.70         0.00             1.9      0.076   
1               7.8              0.88         0.00             2.6      0.098   
2               7.8              0.76         0.04             2.3      0.092   
3              11.2              0.28         0.56             1.9      0.075   
4               7.4              0.70         0.00             1.9      0.076   
...             ...               ...          ...             ...        ...   
6492            6.2              0.21         0.29             1.6      0.039   
6493            6.6              0.32         0.36             8.0      0.047   
6494            6.5              0.24         0.19             1.2      0.041   
6495            5.5              0.29         0.30             1.1      0.022   
6496            6.0              0.21         0.38             0.8      0.020   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                    11.0                  34.0  0.99780  3.51       0.56   
1                    25.0                  67.0  0.99680  3.20       0.68   
2                    15.0                  54.0  0.99700  3.26       0.65   
3                    17.0                  60.0  0.99800  3.16       0.58   
4                    11.0                  34.0  0.99780  3.51       0.56   
...                   ...                   ...      ...   ...        ...   
6492                 24.0                  92.0  0.99114  3.27       0.50   
6493                 57.0                 168.0  0.99490  3.15       0.46   
6494                 30.0                 111.0  0.99254  2.99       0.46   
6495                 20.0                 110.0  0.98869  3.34       0.38   
6496                 22.0                  98.0  0.98941  3.26       0.32   

      alcohol  quality   type  
0         9.4        5    red  
1         9.8        5    red  
2         9.8        5    red  
3         9.8        6    red  
4         9.4        5    red  
...       ...      ...    ...  
6492     11.2        6  white  
6493      9.6        5  white  
6494      9.4        6  white  
6495     12.8        7  white  
6496     11.8        6  white  

[6497 rows x 13 columns]

Before we proceed, let's just convert the column names to camel case

In [6]:
wine_df.columns = [x.replace(' ', '_') for x in wine_df.columns]
wine_df.columns

Index(['fixed_acidity', 'volatile_acidity', 'citric_acid', 'residual_sugar',
       'chlorides', 'free_sulfur_dioxide', 'total_sulfur_dioxide', 'density',
       'pH', 'sulphates', 'alcohol', 'quality', 'type'],
      dtype='object')

## Exploratory data analysis

In [7]:
wine_df.describe().round(3)

fixed_acidity  volatile_acidity  citric_acid  residual_sugar  \
count       6497.000          6497.000     6497.000        6497.000   
mean           7.215             0.340        0.319           5.443   
std            1.296             0.165        0.145           4.758   
min            3.800             0.080        0.000           0.600   
25%            6.400             0.230        0.250           1.800   
50%            7.000             0.290        0.310           3.000   
75%            7.700             0.400        0.390           8.100   
max           15.900             1.580        1.660          65.800   

       chlorides  free_sulfur_dioxide  total_sulfur_dioxide   density  \
count   6497.000             6497.000              6497.000  6497.000   
mean       0.056               30.525               115.745     0.995   
std        0.035               17.749                56.522     0.003   
min        0.009                1.000                 6.000     0.987   
25%        0.038               17.000                77.000     0.992   
50%        0.047               29.000               118.000     0.995   
75%        0.065               41.000               156.000     0.997   
max        0.611              289.000               440.000     1.039   

             pH  sulphates   alcohol   quality  
count  6497.000   6497.000  6497.000  6497.000  
mean      3.219      0.531    10.492     5.818  
std       0.161      0.149     1.193     0.873  
min       2.720      0.220     8.000     3.000  
25%       3.110      0.430     9.500     5.000  
50%       3.210      0.510    10.300     6.000  
75%       3.320      0.600    11.300     6.000  
max       4.010      2.000    14.900     9.000

**Some insights about the data:**<br>

&ensp;- `residual sugar` has a wide range of values [0.600, 65.8]. It might have to do with wines being dry or sweet.    
&ensp;- For the `quality`, there are no values with score lower than 3 and none received the maximum score (10). We could further categotize them.    
&ensp;- Based on common knowledge about wines, the other values seem plausible.

**Wine 'sweetness'**<br>

We can create a column to address the `sweetness` of the wine. According to the [Commission Delegated Regulation (EU) 2019/33](https://eur-lex.europa.eu/legal-content/EN/TXT/PDF/?uri=CELEX:32019R0033&from=en) that regulates the wine sector, applying the terms dry, medium dry, medium sweet or sweet require the following criteria:

**Dry:**<br>
If its sugar content does not exceed:<br>
&ensp;- 4 g$\cdot$dm<sup>-3</sup>, or<br>
&ensp;- 9 g$\cdot$dm<sup>-3</sup>, provided that the total acidity is not more than 2 grams below the residual sugar content.<br>

**Medium dry:**<br>
If its sugar content exceeds the maximum permitted but does not exceed:<br>
&ensp;- 12 g$\cdot$dm<sup>-3</sup>, or<br>
&ensp;- 18 g$\cdot$dm<sup>-3</sup>, provided that the total acidity is not more than 10 grams below the residual sugar content.<br>

**Medium sweet:**<br>
&ensp;If its sugar content exceeds the maximum permitted but does not exceed 45 g$\cdot$dm<sup>-3</sup>.<br>

**Sweet:**<br>
&ensp;If its sugar content is of at least 45 g$\cdot$dm<sup>-3</sup>.<br>

Let's apply these to our data:

In [8]:
def sweet_or_dry(x):
    if (x['residual_sugar'] < 4) | (x['residual_sugar'] < 9 and (x['residual_sugar'] - x['fixed_acidity']) < 2):
        return 'dry'
    elif (x['residual_sugar'] < 12) | (x['residual_sugar'] < 18 and (x['residual_sugar'] - x['fixed_acidity']) < 10):
        return 'medium_dry'
    elif (x['residual_sugar'] < 45):
        return 'medium_sweet'
    else:
        return 'sweet'

In [9]:
wine_df['sweetness'] = wine_df.apply(lambda x: sweet_or_dry(x), axis=1)
wine_df['sweetness'].value_counts(normalize=True)

dry             0.773742
medium_dry      0.200708
medium_sweet    0.025396
sweet           0.000154
Name: sweetness, dtype: float64

The data is totally inbalaced: dry wines correspond to over 77% of the total. Instead of keeping the data like this, we'll only divide them in 'sweet' and 'dry'.

In [10]:
wine_df['sweetness'] = wine_df['sweetness'].apply(lambda x: 'dry' if x == 'dry' else 'sweet')
wine_df['sweetness'].value_counts(normalize=True)

dry      0.773742
sweet    0.226258
Name: sweetness, dtype: float64

**Categorizing the quality score**<br>

We can set an arbitrary threshold to classify the wines into 'low quality' e 'high quality'.<br>
'Low quality': score lower than 6<br>
'High quality': score equal or higher than 6

In [11]:
wine_df['quality_categorized'] = wine_df['quality'].apply(lambda x: 'low' if x < 6 else 'high')
wine_df['quality_categorized'].value_counts(normalize=True)

high    0.633061
low     0.366939
Name: quality_categorized, dtype: float64

In [12]:
wine_df

fixed_acidity  volatile_acidity  citric_acid  residual_sugar  chlorides  \
0               7.4              0.70         0.00             1.9      0.076   
1               7.8              0.88         0.00             2.6      0.098   
2               7.8              0.76         0.04             2.3      0.092   
3              11.2              0.28         0.56             1.9      0.075   
4               7.4              0.70         0.00             1.9      0.076   
...             ...               ...          ...             ...        ...   
6492            6.2              0.21         0.29             1.6      0.039   
6493            6.6              0.32         0.36             8.0      0.047   
6494            6.5              0.24         0.19             1.2      0.041   
6495            5.5              0.29         0.30             1.1      0.022   
6496            6.0              0.21         0.38             0.8      0.020   

      free_sulfur_dioxide  total_sulfur_dioxide  density    pH  sulphates  \
0                    11.0                  34.0  0.99780  3.51       0.56   
1                    25.0                  67.0  0.99680  3.20       0.68   
2                    15.0                  54.0  0.99700  3.26       0.65   
3                    17.0                  60.0  0.99800  3.16       0.58   
4                    11.0                  34.0  0.99780  3.51       0.56   
...                   ...                   ...      ...   ...        ...   
6492                 24.0                  92.0  0.99114  3.27       0.50   
6493                 57.0                 168.0  0.99490  3.15       0.46   
6494                 30.0                 111.0  0.99254  2.99       0.46   
6495                 20.0                 110.0  0.98869  3.34       0.38   
6496                 22.0                  98.0  0.98941  3.26       0.32   

      alcohol  quality   type sweetness quality_categorized  
0         9.4        5    red       dry                 low  
1         9.8        5    red       dry                 low  
2         9.8        5    red       dry                 low  
3         9.8        6    red       dry                high  
4         9.4        5    red       dry                 low  
...       ...      ...    ...       ...                 ...  
6492     11.2        6  white       dry                high  
6493      9.6        5  white       dry                 low  
6494      9.4        6  white       dry                high  
6495     12.8        7  white       dry                high  
6496     11.8        6  white       dry                high  

[6497 rows x 15 columns]

In [13]:
wine_df.to_csv('updated_wine_data.csv')

#### Some questions:
1. Is there a difference in the average quality of red and white wine?
2. What are the variables that influence the quality of the wine? Are they the same for both types?
3. What's the best regression model for the data? Is it best to apply the model to each data separately?
4. More questions?

**Future improvements:**

These datasets can be viewed as classification or regression tasks. The classes are ordered and not balanced (e.g. there are many more normal wines than excellent or poor ones). Outlier detection algorithms could be used to detect the few excellent or poor wines. Also, we are not sure if all input variables are relevant. So it could be interesting to test feature selection methods.

1. Use classification models such as Random Forests
2. Deal with the class imbalancement